tests: bluetooth: classic: Add test suite a2dp_sink

CC0918 opened the pull request:
IUT works as a a2dp sink. The peer device, a2dp source is a PC with running bumble on it.

The test content are follows:
case 1: test API register_ep. The api is used to register endpoint. The case tests it when the parameters is correct or incorrect. case 2: test API a2dp connect. This case repeats the a2dp connect operation to verify its correctness
case 3: test API a2dp disconnect.This case repeats the a2dp connect operation to verify its correctness
case 4: test API a2dp discover. The case send discover and get (all)capabilities cmd to a2dp source.
case 5: This case will complete the execution process of the test a2dp, including connect,discover,get cap, open, start,suspend,close. All cmd are send from source to sink.
case 6: This case will complete the execution process of the test a2dp, including connect,discover,get cap, open, start,suspend,close. All cmd are send from sink to source.

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+
+static void app_disconnected(struct bt_a2dp *a2dp)
+{
+	found_peer_sbc_endpoint = NULL;
```
`default_a2dp` should be cleared here.

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+	}
+
+	type = argv[1];
+	action = argv[2];
```
```suggestion
	codec = argv[2];
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+{
+	uint8_t sbc_hdr;
+
+	if (buf->len < 1U) {
```
```suggestion
	if (buf->len < 7U) {
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+
+static struct bt_a2dp *default_a2dp;
+static uint8_t role;
+static uint8_t a2dp_initied;
```
```suggestion
static bool a2dp_initied;
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+		}
+	} else {
+		shell_error(sh, "a2dp is not connected");
+	}
```
if (default_a2dp == NULL) {
		shell_error(sh, "a2dp is not connected");
	}

	err = bt_a2dp_disconnect(default_a2dp);
	if (err == 0) {
		default_a2dp = NULL;
	} else {
		shell_print(sh, "fail to send disconnect cmd");
	}

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+			uint32_t ts)
+{
+	sink_sbc_streamer_data(stream, buf, seq_num, ts);
+}
```
```suggestion
#if defined(CONFIG_BT_A2DP_SINK)
static void stream_recv(struct bt_a2dp_stream *stream, struct net_buf *buf, uint16_t seq_num,
			uint32_t ts)
{
	sink_sbc_streamer_data(stream, buf, seq_num, ts);
}
#endif /* CONFIG_BT_A2DP_SINK */
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+		       (uint8_t)BT_A2DP_SBC_MEDIA_HDR_NUM_FRAMES_GET(sbc_hdr), buf->len);
+	bt_shell_print("data: %d, %d, %d, %d, %d, %d ......", buf->data[0], buf->data[1],
+		       buf->data[2], buf->data[3], buf->data[4], buf->data[5]);
+}
```
```suggestion
#if defined(CONFIG_BT_A2DP_SINK)
static void sink_sbc_streamer_data(struct bt_a2dp_stream *stream, struct net_buf *buf,
				   uint16_t seq_num, uint32_t ts)
{
	uint8_t sbc_hdr;

	if (buf->len < 1U) {
		return;
	}
	sbc_hdr = net_buf_pull_u8(buf);
	bt_shell_print("received, num of frames: %d, data length:%d",
		       (uint8_t)BT_A2DP_SBC_MEDIA_HDR_NUM_FRAMES_GET(sbc_hdr), buf->len);
	bt_shell_print("data: %d, %d, %d, %d, %d, %d ......", buf->data[0], buf->data[1],
		       buf->data[2], buf->data[3], buf->data[4], buf->data[5]);
}
endif /* CONFIG_BT_A2DP_SINK */
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+		}
+	} else {
+		shell_error(sh, "a2dp is not connected");
+	}
```
```suggestion
	if (default_a2dp == NULL) {
		shell_error(sh, "a2dp is not connected");
		return -ENOEXEC;
	}

	if (registered_sbc_endpoint == NULL) {
		shell_error(sh, "no endpoint");
		return -ENOEXEC;
	}

	if (found_peer_sbc_endpoint == NULL) {
		shell_error(sh, "don't find the peer sbc endpoint");
		return -ENOEXEC;
	}

	bt_a2dp_stream_cb_register(&sbc_stream, &stream_ops);

	err = bt_a2dp_stream_config(default_a2dp, &sbc_stream, registered_sbc_endpoint,
					found_peer_sbc_endpoint, &sbc_cfg);
	if (err != 0) {
		shell_error(sh, "fail to configure (err %d)", err);
		return -ENOEXEC;
	}
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+		}
+	} else {
+		bt_shell_error("fail to get avdtp version");
+	}
```
```suggestion
	if (err != 0) {
		bt_shell_error("fail to get avdtp version");
		return BT_SDP_DISCOVER_UUID_CONTINUE;
	}

	if (default_a2dp != NULL) {
		discover_param.avdtp_version = peer_avdtp_version;
		err = bt_a2dp_discover(default_a2dp, &discover_param);
		if (err != 0) {
			bt_shell_error("fail to discover peer endpoints");
		}
	}
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,745 @@
+	SHELL_CMD_ARG(release, NULL, "\"release the stream\"", cmd_release, 1, 0),
+	SHELL_CMD_ARG(start, NULL, "\"start the stream\"", cmd_start, 1, 0),
+	SHELL_CMD_ARG(suspend, NULL, "\"suspend the stream\"", cmd_suspend, 1, 0),
+	SHELL_CMD_ARG(abort, NULL, "\"abort the stream\"", cmd_abort, 1, 0), SHELL_SUBCMD_SET_END);
```
```suggestion
	SHELL_CMD_ARG(configure, NULL, HELP_NONE, cmd_configure, 1, 0),
	SHELL_CMD_ARG(establish, NULL, HELP_NONE, cmd_establish, 1, 0),
	SHELL_CMD_ARG(reconfigure, NULL, HELP_NONE, cmd_reconfigure, 1, 0),
	SHELL_CMD_ARG(release, NULL, HELP_NONE, cmd_release, 1, 0),
	SHELL_CMD_ARG(start, NULL, HELP_NONE, cmd_start, 1, 0),
	SHELL_CMD_ARG(suspend, NULL, HELP_NONE, cmd_suspend, 1, 0),
	SHELL_CMD_ARG(abort, NULL, HELP_NONE, cmd_abort, 1, 0), 
	SHELL_SUBCMD_SET_END
);
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,749 @@
+	if (rsp_err_code == 0) {
+		bt_shell_print("success to establish");
+	} else {
+		bt_shell_print("fail to establish");
```
```suggestion
		bt_shell_error("fail to establish");
```

CC0918 commented on tests/bluetooth/classic/a2dp_sink/pytest/test_a2dp.py:
```diff
@@ -0,0 +1,726 @@
+
+        device.host.snooper = BtSnooper(snoop_file)
+        await device_power_on(device)
+        await device.send_command(HCI_Write_Page_Timeout_Command(page_timeout=0xFFFF))
```
for furture to case by case test

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,774 @@
+	bt_shell_print("stream suspended");
+}
+
+#if defined(CONFIG_BT_A2DP_SINK)
```
I think the pre-condition could be removed.

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,774 @@
+	} else {
+		shell_error(sh, "A2DP endpoint role not set or invalid");
+		return -EINVAL;
+	}
```
Peer device should be a A2DP source here. I think the code block could be optimized.

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,902 @@
+		shell_help(sh);
+		return 0;
+	}
+
```
```suggestion
	if (err != 0) {
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,792 @@
+		}
+	}
+	return 0;
+}
```
```suggestion
		shell_error(sh, "fail");
```

lylezhu2012 commented on tests/bluetooth/classic/a2dp_sink/src/test_a2dp_sink.c:
```diff
@@ -0,0 +1,792 @@
+	"min_b: 2-250\n"                                                                           \
+	"max_b: 2-250\n"                                                                           \
+	">"
+
```
```suggestion
		shell_error(sh, "fail");
```

CC0918 committed:
tests: bluetooth: classic: Add test suite a2dp_sink

IUT works as a a2dp sink. The peer device, a2dp source is
a PC with running bumble on it.

The test content are follows:
case 1: test API register_ep. The api is used to register endpoint.
The case tests it when the parameters is correct or incorrect.
case 2: test API a2dp connect. This case repeats the a2dp connect
operation to verify its correctness
case 3: test API a2dp disconnect.This case repeats the a2dp connect
operation to verify its correctness
case 4: test API a2dp discover. The case send discover and get
(all)capabilities cmd to a2dp source.
case 5: This case will complete the execution process of the test a2dp,
including connect,discover,get cap, open, start,suspend,close.
All cmd are send from source to sink.
case 6: This case will complete the execution process of the test a2dp,
including connect,discover,get cap, open, start,suspend,close.
All cmd are send from sink to source.

Signed-off-by: Cheng Chang <cheng.chang@nxp.com>
